bindings/rust/s2n-tls-hyper/src/connector.rs

```diff
@@ -118,7 +126,11 @@ where
             return Box::pin(async move { Err(Error::InvalidScheme) });
         }
 
-        let builder = self.conn_builder.clone();
+        // Attempt to negotiate HTTP/2.
+        let builder = connection::ModifiedBuilder::new(self.conn_builder.clone(), |conn| {
+            conn.set_application_protocol_preference([b"h2"])
+        });
```

Comment:
"set_application_protocol_preference" erases any existing protocol preferences. That seems a little unexpected to me-- is there any reason an application would have a protocol other than "h2" in there? Should you maybe be doing "append_application_protocol_preference" instead?

Reply:
I agree that erasing existing protocol preferences could be unexpected, however I think appending h2 to existing preferences could be unexpected as well. If someone really cares about the application protocol that s2n-tls-hyper negotiates, it seems wrong to append additional options to their preferences. Instead, if customizing the HTTP values in the ALPN is someday required, I think we should provide a way to configure this in s2n-tls-hyper.

> is there any reason an application would have a protocol other than "h2" in there?

I'm not aware of a reason to do this, other than an application not wanting to negotiate HTTP/2.

Since s2n-tls-hyper is the component negotiating the application protocol (HTTP), it seems like s2n-tls-hyper should be the one to decide what is included in the ALPN, even if this involves adding configuration options to s2n-tls-hyper like the ability to not negotiate HTTP/2.

Reply:
Idk, I guess my thinking is that if we don't expect any other protocols in there but don't specifically need zero other protocols in there, appending would be the more conservative choice? But I can see both arguments. Is either option a one-way door?

Reply:
Hmm yeah I guess that makes sense as well.

If we assume there's a legitimate reason someone might want to set their own protocol preferences in addition to h2, then I suppose both options are one-way, since we'd change how the provided preferences are used in both cases. However, I don't currently see how this would impact anything (while still producing a functioning HTTP connection), so I think both options are probably two-way.

I'll change it to appending for now if you think that'd be less unexpected.

Reply:
> appending would be the more conservative choice?

Hmm. I worry that appending is sort of "breaking" s2n-tls-hyper expectations? `s2n-tls-hyper` relies on being able to control the ALPN, and a customer setting their own ALPN is sort of "tampering" with that behavior?

But I do agree that erasing existing protocol preferences is a bit unexpected. Is there any way to return an error? A customer setting ALPNs means that they want to do ALPN things, but _we_ want to do ALPN things, and it's unclear how those will interact.

Reply:
Yeah that makes sense. I agree that s2n-tls-hyper should ideally be the only one setting the ALPN.

Erroring could work, but unfortunately we can't know in `HttpsConnector::new` whether an ALPN was set, since it's only provided a `connection::Builder`, which doesn't allow access to the `Config` (and the builder may not even be a `Config`). We could maybe add a `alpn_was_set` API to the connection, and error in `Service::call` when making the HTTP request, but erroring here seems late and could be unexpected. I'll look into this more!

Reply:
We discussed this offline. We decided it was best to have s2n-tls-hyper be the only one modifying the ALPN extension, so overriding existing values was preferred over appending.

We also decided it was best to override rather than error, since the error would have to occur when sending the HTTP request rather than during HttpsConnector construction, unless HttpsConnector was modified to be generic over Config rather than connection::Builder.